### DISCUSSION TITLE
Dependency injection exceptions

### DISCUSSION THREAD

**Discussion by ADarko22** [micronaut-projects/micronaut-core#10372] [General]
Hello, 

I have a question about the *dependency injection* in Micronaut.

According to the [documentation on injection types](https://guides.micronaut.io/latest/micronaut-dependency-injection-types-maven-java.html), three types are supported: constructor, field, and method parameter. Anyway, it seems there are some exceptions in which not all types are valid. For instance, the constructor injection appears to not work for  **MicronautRequestHandler**, in fact, the documentation on [Micronaut AWS](https://micronaut-projects.github.io/micronaut-aws/1.1.0.M1/guide/index.html) only mentions field injection.

Can you confirm that this is the case? In that case, are there other exceptions to be aware of? 
And, if those exceptions exist, is there any plan to support the constructor injections for them?

Thank you,
Angelo

**Comment by Jonathan-Zollinger** [comment]
> For instance, the constructor injection appears to not work for MicronautRequestHandler

why does this not work? to help me understand, is this an assumption based on the mn-aws docs doesn't include construction injection or do you have some code that isn't behaving as expected?

**Reply by marcin-majewski-sonarsource** [reply]
> why does this not work? to help me understand, is this an assumption based on the mn-aws docs doesn't include construction injection or do you have some code that isn't behaving as expected?

To be more precise, it does not work due to the fact that empty constructor is required in the handler class by AWS. So this is not valid handler for AWS:
```java
public class SampleHandler extends MicronautRequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {

  private Environment environment;

  public SampleHandler(Environment environment) {
    this.environment = environment;
  }

  @Override
  public APIGatewayProxyResponseEvent execute(APIGatewayProxyRequestEvent input) {
    return new APIGatewayProxyResponseEvent()
      .withStatusCode(200)
      .withBody("Hello World");
  }
}
```

hence field injection is the only option. I think the question here (as I discussed offline with @ADarko22) is whether this is expected or is there another way to utilize constructor injection here?
